### PR TITLE
Add support for metalsmith builder hooks

### LINKED ISSUE #427: Custom metalsmith plugin
I need to use a custom metalsmith plugin in after default vue-cli ones. (see [generate.js#L47-L49](https://github.com/vuejs/vue-cli/blob/439aece38f54c32f7e0de29f483f24f6523e7a80/lib/generate.js#L47-L49))

In my case, I need to move some files to another destination based on answers data, but I think it would be great to let writers of custom templates allow adding any metalsmith plugin in the chain.

### PR DESCRIPTION
**Pull request by Toilal**
Close #427

### CODE REVIEW

**Comment by zigomir** [comment]
I think this would be even better with tests.

**Comment by Toilal** [comment]
I'll write one test that writes target files to another destination, as I think it would be a common use case.

**Comment by Toilal** [comment]
Test written. I have also removed `it.only` in `test.js` that was causing only one test to run.

File: test/e2e/test.js [hunk 1]
```diff
@@ -254,7 +265,7 @@ describe('vue-cli', () => {
     expect(getTemplatePath('../template')).to.equal(path.join(__dirname, '/../../../template'))
   })
 
-  it.only('points out the file in the error', done => {
```
Comment (zigomir) [inline test/e2e/test.js]: 👍

File: test/e2e/mock-metalsmith-custom/meta.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  "metalsmith": function (metalsmith, opts, helpers) {
+    function customMetalsmithPlugin (files, metalsmith, done) {
+      // Implement something really custom here.
+
+      var readme = files['readme.md']
+      delete files['readme.md']
+      files['custom/readme.md'] = readme
+
+      console.log(files)
```
Comment (zigomir) [inline test/e2e/mock-metalsmith-custom/meta.js]: I believe this could be removed.

File: test/e2e/mock-metalsmith-custom/meta.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  "metalsmith": function (metalsmith, opts, helpers) {
+    function customMetalsmithPlugin (files, metalsmith, done) {
+      // Implement something really custom here.
+
+      var readme = files['readme.md']
+      delete files['readme.md']
+      files['custom/readme.md'] = readme
```
Comment (zigomir) [inline test/e2e/mock-metalsmith-custom/meta.js]: Just to be sure. What you're doing here in custom metalsmith plugin is you're moving file from root template directory into `custom` directory, right?

File: test/e2e/mock-metalsmith-custom/meta.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  "metalsmith": function (metalsmith, opts, helpers) {
```
Comment (zigomir) [inline test/e2e/mock-metalsmith-custom/meta.js]: Besides this one you can also test `before` and `after`, right?

File: test/e2e/mock-metalsmith-custom/meta.js (hunk 2 above)
Comment (Toilal) [inline test/e2e/mock-metalsmith-custom/meta.js]: oups yes :)

File: test/e2e/mock-metalsmith-custom/meta.js (hunk 3 above)
Comment (Toilal) [inline test/e2e/mock-metalsmith-custom/meta.js]: Yes that's it.

File: test/e2e/mock-metalsmith-custom/meta.js (hunk 4 above)
Comment (Toilal) [inline test/e2e/mock-metalsmith-custom/meta.js]: ok

Comment (Toilal) [inline test/e2e/mock-metalsmith-custom/meta.js]: I just push a new commit with an additional test, and better assertions

**Comment by zigomir** [comment]
@Toilal looks good, thank you!

**Pull request merged**